### PR TITLE
Improve/validator transform improvements

### PR DESCRIPTION
**Pull request by gsps1**
validator transform improvements - https://issues.cask.co/browse/CDAP-3612

### CODE REVIEW

**Comment by gsps1** [comment]
http://builds.cask.co/browse/CDAP-RBT419-1

**Comment by elfenheart** [comment]
@gsps1 Is there a UI change needed for this?

**Comment by gsps1** [comment]
@elfenheart no, these are just backend improvements

**Comment by elfenheart** [comment]
@gsps1 But the script that the UI is sending to the backend, it won't be java hashmap and we don't need to JSON.parse the input anymore? Those codes are getting constructed in the UI as well

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java [hunk 1]
```diff
@@ -127,6 +121,8 @@ void setUpInitialScript(TransformContext context, List<Validator> validators) th
     for (Validator validator : validators) {
       engine.put(validator.getValidatorName(), validator.getValidator());
     }
+    String jsonParseScript = "function parseJSON(input) { return JSON.parse(input); }";
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java]: should do this the same way ScriptTransform does it.  I don't remember the exact reason, but there was some issue with JSON.parse().  I think it's not always available depending on the environment or something like that.

File: cdap-docs/application-templates/source/etl/templates/transformations/validator.rst [hunk 2]
```diff
@@ -26,16 +26,15 @@ valid.
 Example: ``"validators": "core"``
 
 **validationScript:** Javascript that implements the function ``isValid``, taking a JSON object
-representation of the input record, and returning a ``Map<String, String>`` representing the result.
-The map should have these fields::
+representation of the input record, and returning a JSON representing the result.
+The result JSON should have these fields, errorCode and errorMsg can be ignored for valid records::
```
Comment (awholegunch) [inline cdap-docs/application-templates/source/etl/templates/transformations/validator.rst]: The returned JSON will include these fields;  ``errorCode`` and ``errorMsg`` can be ignored for valid records::

File: cdap-docs/application-templates/source/etl/templates/transformations/validator.rst [hunk 3]
```diff
@@ -26,16 +26,15 @@ valid.
...
 
   {
     "isValid" : "true [or] false",
-    "errorCode" : "number",
+    "errorCode" : "number [should be an valid integer] ",
```
Comment (awholegunch) [inline cdap-docs/application-templates/source/etl/templates/transformations/validator.rst]: "errorCode" : "number [will be a valid integer] ",

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java [hunk 4]
```diff
@@ -66,8 +65,6 @@
     "      var isValid = true; " +
```
Comment (awholegunch) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java]: The String SCRIPT_DESCRIPTION needs to be updated to reflect the changes in the RST file.

**Comment by awholegunch** [comment]
I can look at this again after the code is updated.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java [hunk 5]
```diff
@@ -135,41 +131,37 @@ void setUpInitialScript(TransformContext context, List<Validator> validators) th
...
       }
     } catch (Exception e) {
       throw new IllegalArgumentException("Invalid filter condition.", e);
     }
   }
 
-  private InvalidEntry<StructuredRecord> getErrorObject(Map<String, String> result, StructuredRecord input) {
+  private InvalidEntry<StructuredRecord> getErrorObject(Map result, StructuredRecord input) {
     Preconditions.checkState(result.containsKey("errorCode"));
     Preconditions.checkState(result.containsKey("errorMsg"));
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java]: we should allow empty error msg

**Comment by gsps1** [comment]
@elfenheart you meant that, yeah. that config needs to be updated.

File: cdap-docs/application-templates/source/etl/templates/transformations/validator.rst [hunk 6]
```diff
@@ -26,16 +26,15 @@ valid.
 Example: ``"validators": "core"``
 
 **validationScript:** Javascript that implements the function ``isValid``, taking a JSON object
-representation of the input record, and returning a ``Map<String, String>`` representing the result.
-The map should have these fields::
+representation of the input record, and returning a JSON representing the result.
+The result JSON should have these fields, errorCode and errorMsg can be ignored for valid records::
 
   {
     "isValid" : "true [or] false",
```
Comment (albertshau) [inline cdap-docs/application-templates/source/etl/templates/transformations/validator.rst]: don't put quotes, it makes it look like it should be a string

File: cdap-docs/application-templates/source/etl/templates/transformations/validator.rst (hunk 3 above)
Comment (albertshau) [inline cdap-docs/application-templates/source/etl/templates/transformations/validator.rst]: don't put quotes, it makes it look like it should be a string

Comment (gsps1) [inline cdap-docs/application-templates/source/etl/templates/transformations/validator.rst]: should makes more sense? as we fail if the error code is not a valid integer.

**Comment by gsps1** [comment]
@awholegunch @albertshau have addressed the comments.

**Comment by albertshau** [comment]
one comment, otherwise code lgtm

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java [hunk 7]
```diff
@@ -58,7 +57,7 @@
 
   private static final String SCRIPT_DESCRIPTION = "Javascript that must implement a function 'isValid' that " +
     "takes a JSON object representation of the input record, " +
-    "and returns as a result a Map<String, String>, " +
+    "and returns a result JSON " +
```
Comment (awholegunch) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java]: "and returns a result JSON. " +

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java [hunk 8]
```diff
@@ -58,7 +57,7 @@
...
     "Example response : " +
```
Comment (awholegunch) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java]: "Example response: " +

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java [hunk 9]
```diff
@@ -120,14 +117,21 @@ public void initialize(TransformContext context) throws Exception {
...
 
     for (Validator validator : validators) {
       engine.put(validator.getValidatorName(), validator.getValidator());
     }
-    engine.eval(scriptStr);
+
+    // this is pretty ugly, but doing this so that we can pass the 'input' json into the isValid function.
+    // that is, we want people to implement
+    // function isValid(input) { ... }
+    // rather than function isValid() { ... } and have them access a global variable in the function
```
Comment (awholegunch) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java]: and have them be able to access global variables in the function [is that a correct description?]

**Comment by awholegunch** [comment]
Three minor comments, otherwise LGTM.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java (hunk 9 above)
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java]: the existing one is clearer i think.

Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/ValidatorTransform.java]: added little more details

**Comment by gsps1** [comment]
i will merge after the build, missed updating the script in the earlier build.

**Comment by gsps1** [comment]
Build green, merging.

**Pull request merged**